Refactor Save Options Handling

DavidFair opened the pull request:
**Description of work.**
Refactors the Save Options handling in the ISIS SANS GUI to correctly follow our GUI standards.
This is prep work for #31929 

**To test:**
- Download the ISIS Training Data and ensure the loqdemo folder is present
- Open the ISIS SANS GUI
- Set the mask file to the .TOML file in the loqdemo folder
- Set the batch file to the .csv from the same folder
- Under manage user directories change the default save dir to somewhere clean
- Set the output mode to file 
- Mess around with the tick options and check they work. Note the default options missing is expected, as it's being done in an upcoming PR

Part of #31929 

*This does not require release notes* because this partially completes the issue

<!-- Ensure the base of this PR is correct (e.g. release-next or master)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DavidFair committed:
Add initial generic observable handler for save opts

This adds a obserable handler for save options in a way that we can
extend it in the future.

It also starts introducing a basic MVP and observer pattern to the GUI,
as per the coding standards

DavidFair committed:
Forward on selected types from view

Sets up the presenter to retrieve the save types from the view and
forward them onto the model

DavidFair committed:
Add handling for parsing SaveTypes

Adds handling to parse SaveTypes from the view layer in the model

DavidFair committed:
Pull update from model instead of view in SANS

Pull the current save options from the model instead of the View in ISIS
SANS

DavidFair committed:
Mirror handling in model to logic in view

Mirror the logic in the view into the save options model

DavidFair committed:
Switch to using SaveOptions in the view and run_tab_presenter

Switches to using simpler SaveOptions which is a POD type rather than a
list in the view

DavidFair committed:
Remove print statement from model for diag purposes

This was useful to ensure the signal / slot pair was wired up correctly,
but is too noisy

DavidFair committed:
Avoid error on older Mock where __round__ is not defined

Avoids an error where we implicitly call __round__ which is not defined
for older Mock packages, but locally works fine

Pasarus reviewed (changes requested):
One small question other than that it looks good! :+1:

DavidFair committed:
Rename SANS model tests to match test_* convention

Brings the SANS model tests in-line with the project. SANS tests were
previously mixed as some pre-dated the coding standard which specifies
test_*

DavidFair commented on scripts/test/SANS/gui_logic/model_tests/CMakeLists.txt:
```diff
@@ -1,6 +1,8 @@
-                  RowEntriesTest.py
-                  settings_adjustment_model_test.py
+set(TEST_PY_FILES
+    file_loading_test.py
```
Done, I've not changed the names of the classes but I'll bring them to read TestXYZ as I work with them in the future

Good spot :+1:
